### ISSUE TITLE
NPU runtime: mixed GPU→NPU multi-stage pipeline corrupts output (NPU relay downstream of GPU)

### ISSUE THREAD

**Issue by t8** [labscommunity/cascadia#67]
## Summary

In a **heterogeneous multi-stage pipeline**, an **NPU stage placed as a relay
downstream of a GPU stage produces corrupt output** (degenerate garbage, 0/12
on the eval suite). A single NPU stage — even the embedding stage, even feeding
GPU downstream — is correct; the corruption appears only when an NPU stage is a
*non-first relay* with non-NPU stages upstream of it.

This is an **NPU static-KV runtime** gap (the multi-stage path from #63), not a
placement-logic problem. It was surfaced while benchmarking #41/#66's
`run-placement` with a deliberately-suboptimal "offload to the NPU" baseline.
It does **not** affect #41/#66: the placement ILP measures the NPU as the
slowest tier for decode (GPU < CPU < NPU) and therefore *never* assigns a stage
to the NPU — every placement the solver emits (GPU-only, GPU+CPU) runs correct.
This is filed so the underlying mixed-pipeline NPU bug isn't lost.

Related: #37 (NPU op-support), #63 (NPU static-shape + multi-stage runtime), #66 (placement).

## Evidence

All runs: static export (`cascadia shard --target npu --quantization fp16`),
greedy decode, 12-prompt eval suite, on pawan-01 (Lunar Lake, Arc 140V iGPU +
AI Boost NPU), via the `ov-runtime` engine.

| Model | placement (per-stage device) | topology | verdict |
|---|---|---|---|
| Yi-1.5-9B fp16 (12 stages) | `NPU, GPU×11` | NPU(0)→GPU(1..11) | **12/12 PASS**, `corrupt=False` |
| SOLAR-10.7B fp16 (12 stages) | `NPU, GPU×9, NPU, NPU` | NPU(0)→GPU(1..9)→NPU(10)→NPU(11=head) | **0/12 CORRUPT** |

For contrast, **every homogeneous or GPU/CPU-only placement of the same shards
is correct** (12/12): all-GPU, all-NPU-first→GPU, and the ILP's GPU+CPU
placements on Qwen2.5-32B / Yi-9B / SOLAR all pass. So the corruption is
specific to an NPU stage acting as a **relay whose upstream chain includes a
non-NPU (GPU) stage**.

What changed between the PASS and CORRUPT rows (the suspects):
1. a **GPU→NPU mid-pipeline handoff** (GPU stage 9 → NPU stage 10) — this
   *direction* is new; #63 only validated NPU→NPU and NPU-first→GPU;
2. an **NPU relay stage whose upstream is GPU** (stage 10);
3. the **`lm_head` final stage on the NPU** (stage 11).

## Repro

```bash
# 1. Export a small static fp16 shard with >=3 stages (fast; reproduces without a big model):
cascadia shard --model Qwen/Qwen2.5-1.5B-Instruct \
  --output-dir /tmp/q15-fp16-3stage --num-stages 3 \
  --target npu --static-context 256 --quantization fp16

# 2. Launch a heterogeneous ring with an NPU relay DOWNSTREAM of a GPU stage,
#    i.e. rank0=GPU (first), rank1=NPU (relay after GPU), rank2=NPU/GPU (last):
cascadia worker --rank 0 --total 3 --engine ov-runtime --device GPU \
  --model /tmp/q15-fp16-3stage --next 127.0.0.1:9101 --api 127.0.0.1:8000 &
cascadia worker --rank 1 --total 3 --engine ov-runtime --device NPU \
  --model /tmp/q15-fp16-3stage --listen 127.0.0.1:9101 --next 127.0.0.1:9102 &
cascadia worker --rank 2 --total 3 --engine ov-runtime --device NPU \
  --model /tmp/q15-fp16-3stage --listen 127.0.0.1:9102 &

# 3. Compare against the KNOWN-GOOD control (NPU only as the FIRST stage):
#    rank0=NPU, rank1=GPU, rank2=GPU  -> expected correct.

# 4. Hit the API; the GPU-first/NPU-relay ring emits garbage, the NPU-first ring is coherent.
curl -s 127.0.0.1:8000/v1/chat/completions -d '{"messages":[{"role":"user","content":"What is the capital of France?"}],"max_tokens":16}'
```

(#66's `cascadia run-placement` + a hand-written `placement.json` automates the
launch, but the bug reproduces on current `main` with the manual ring above.)

## Likely root cause

The #63 static-KV multi-stage design (`crates/cascadia-engine-openvino/src/runtime.rs`)
carries the absolute `position` from the **first** stage as its own framed wire
tensor, while **relay** stages derive position from their host-side KV-ring
counter (reset to 0, incremented per token). This was validated for the
homogeneous NPU→NPU ring and for NPU-first→GPU. The mixed case where an NPU
**relay** sits downstream of GPU stages is unvalidated, and the most likely
failure is that the position/RoPE state is not propagated/consumed consistently
across the GPU→NPU boundary (the downstream NPU relay attends at the wrong
absolute position → garbled attention → degenerate output). The `lm_head` on
the NPU is a secondary suspect.

## Steps to fix

**1. Isolate the exact transition** (minimal 3-stage static fp16 shard; one
variable at a time):
- `GPU(0) → NPU(1) → GPU(2)` — NPU as a *middle* relay after GPU (no head on NPU). Isolates the GPU→NPU relay handoff.
- `GPU(0) → GPU(1) → NPU(2=head)` — NPU only as the head. Isolates head-on-NPU.
- `NPU(0) → GPU(1) → NPU(2)` — NPU first + NPU relay-after-GPU.
Whichever corrupts pins the cause (vs. the known-good `NPU(0)→GPU(1)→GPU(2)`).

**2. Fix in the static-KV runtime** (`runtime.rs`, the `run_first`/`run_relay`
+ `send_hidden_downstream`/`recv_hidden_from_upstream` +
`encode_wire_position`/`decode_wire_position` paths):
- Make the absolute `position` propagate through **every** stage boundary, not
  just first→second, so a downstream NPU relay gets the correct position
  regardless of the upstream device (today only the first stage frames it).
  Either forward the position frame end-to-end, or have each relay carry its
  position downstream the same way the first stage does.
- Verify the hidden-states dtype/layout is identical across a GPU→NPU handoff
  (the framed wire tensor) — confirm no fp16/precision or shape-rank mismatch.
- If the head-on-NPU path is implicated, verify the last-stage logits handoff
  works when the final stage is NPU in a heterogeneous ring.

**3. Guard + validate:**
- Add a mixed GPU+NPU multi-stage e2e test (and an int4 variant — only fp16 was
  observed; int4 mixed-NPU-relay is untested) to the suite, asserting `12/12`.
- Until fixed, optionally have the worker/`run-placement` **warn** (or refuse)
  when a placement puts a non-first NPU stage downstream of a non-NPU stage,
  pointing at this issue.

## Priority

Low-to-medium. The placement ILP (#66) never selects an NPU stage for decode
(it's the slowest tier), so no tool-generated placement hits this. It matters
if/when the NPU becomes worth using in a mixed pipeline (e.g. for a future
throughput/overlap regime, or a model class where the NPU is competitive).

**Comment by t8** [comment]
## Correction: this is NOT a corruption bug (first-principles diagnosis)

Before fixing, I reproduced from scratch on a small model and the original premise does **not** hold.

**Diagnostic — Qwen2.5-1.5B, 3-stage static, every NPU topology PASSES:**

| config | verdict |
|---|---|
| fp16 `NPU,GPU,GPU` (NPU first) | 12/12 |
| fp16 `NPU,NPU,NPU` (all NPU) | 12/12 |
| fp16 `GPU,NPU,GPU` (NPU **middle relay after GPU** — the suspected bug) | **12/12** |
| fp16 `GPU,GPU,NPU` (NPU head after GPU) | 12/12 |
| int4 `GPU,NPU,GPU` / `NPU,NPU,NPU` | 11/12 (tiny-model miss, `corrupt=False`) |

So a mixed GPU→NPU relay does **not** corrupt.

**What actually happened in the SOLAR run:** the "corrupt" eval results were all
`out=""` at exactly `secs=60.07` — i.e. every request hit the eval's **timeout**,
not garbled tokens. That config placed **3 stages of a 10.7B model on the NPU**
(the slowest tier for decode), running at **0.14 tok/s** (~7 s/token). At that
speed every eval request times out → empty response → the eval's degeneracy
detector flags empty-string as `corrupt=True` (a **false positive**).

Re-running the same SOLAR placement with a long client timeout produces coherent
output: *"Paris, the captivating capital of France, is renowned for its romantic
ambiance, world-class art, and rich cultural…"* — **correct, just slow.**

**Conclusion:** the NPU static-KV multi-stage runtime is correct on mixed
pipelines; there is no output corruption. The only real (minor) takeaway is that
the eval harness mislabels an empty/timeout response as `corrupt`. Recommend
**closing this as not-a-bug** (or repointing it at the eval-harness false
positive). The ILP already never places decode stages on the NPU, so the slow
config isn't reachable via the tool anyway.

**Comment by t8** [comment]
Fix up in #68 — but note the **re-scope**: this was *not* a corruption bug. Diagnosis (in the comment above) showed the runtime is correct on every NPU topology; the slow/'corrupt' SOLAR run was RAM exhaustion (20 GiB model across 12 workers → ~0 GiB free → swap) with the eval mislabeling timed-out empty responses as corrupt. #68 makes `cascadia place` warn when a placement's footprint will exhaust RAM, which is the real, actionable fix.
